proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +267,232 @@ protocol TargetBuildContext {
 
 It is important to emphasize the distinction between a package plugin and any build commands it defines:
 
-- The *plugin* is invoked after the structure of the target that uses it is known but before any build commands are run. It may also be invoked again if there the target structure or other input conditions change. It is not invoked again when the *contents* of files change.
+- The *plugin* is invoked after the structure of the target that uses it is known, but before any build commands are run. It may also be invoked again if there the target structure or other input conditions change. It is not invoked again when the *contents* of files change.
 
-- The *commands* that are constructed by an plugin using the `CommandConstructor` API are invoked before, during, or after each build, in accordance with their defined input and output dependencies.
+- The *commands* that are created by an plugin using the `CommandConstructor` API are invoked either before or during the build, in accordance with their defined input and output dependencies.
 
 Binary targets will be improved to allow them to contain command line tools and required support files (such as the system `.proto` files in the case of `protoc`, etc). Binary targets will need to support different executable binaries based on platform and architecture. This is the topic of a separate evolution proposal to extend binary targets to support executables in addition to XCFrameworks.
 
 #### Plugin Capabilities
 
-This proposal defines three different plugin capabilities, which determine how the plugin is used.
+Althought this proposal defines a general approach to plugins, it defines only one initial capability: `buildTool`.  This capability is used for both prebuild commands and build commands, and may in the future be used for additional kinds of build tool invocations (in which varations in build commands would be expressed through new `CommandConstructor` APIs).  Significantly different types of plugin functionality are likely to use different capabilities.
+
+In the package manifest, the capability is expressed a function invocation rather than an enum in order to make it easier to add arguments with default values in the future.
+
+#### Prebuild Command Semantics
+
+Commands created using `addPrebuildCommand()` are run before the start of every build. When creating prebuild commands, the plugin needs to specify a directory into which the outputs will be written.  This is how the prebuild command communicates its outputs to the build system.
+
+Before invoking the prebuild command, the build system will create the output directory if needed, but it will not remove any directory contents that already exist.  After invoking the command, SwiftPM will use the contents of that directory as inputs to the consruction of build commands. The prebuild command should add or remove files so that the directory contents match the source files that should be processed by the build system.
```

Comment:
```suggestion
Before invoking the prebuild command, the build system will create the output directory if needed, but it will not remove any directory contents that already exist.  After invoking the command, SwiftPM will use the contents of that directory as inputs to the construction of build commands. The prebuild command should add or remove files so that the directory contents match the source files that should be processed by the build system.
```